README.md
# node-goldwasher-needle
[![npm version](http://img.shields.io/npm/v/goldwasher-needle.svg)](https://www.npmjs.org/package/goldwasher-needle)
[![Build Status](http://img.shields.io/travis/alexlangberg/node-goldwasher-needle.svg)](https://travis-ci.org/alexlangberg/node-goldwasher-needle)
[![Coverage Status](http://img.shields.io/coveralls/alexlangberg/node-goldwasher-needle.svg)](https://coveralls.io/r/alexlangberg/node-goldwasher-needle?branch=master)
[![Code Climate](http://img.shields.io/codeclimate/github/alexlangberg/node-goldwasher-needle.svg)](https://codeclimate.com/github/alexlangberg/node-goldwasher-needle)

[![Dependency Status](https://david-dm.org/alexlangberg/node-goldwasher-needle.svg)](https://david-dm.org/alexlangberg/node-goldwasher-needle)
[![devDependency Status](https://david-dm.org/alexlangberg/node-goldwasher-needle/dev-status.svg)](https://david-dm.org/alexlangberg/node-goldwasher-needle#info=devDependencies)
[![peerDependency Status](https://david-dm.org/alexlangberg/node-goldwasher-needle/peer-status.svg)](https://david-dm.org/alexlangberg/node-goldwasher-needle#info=peerDependencies)

Plugin for [goldwasher](https://www.npmjs.org/package/goldwasher) to add [needle](https://www.npmjs.org/package/needle) for easy HTTP requests. Uses [retry](https://www.npmjs.org/package/retry) for robustness. Requires [goldwasher](https://www.npmjs.org/package/goldwasher) to work. If you want to run this as a web server/service, have a look at the module [hapi-goldwasher](https://www.npmjs.com/package/hapi-goldwasher).

## Installation
```
npm install goldwasher-needle
```

## Options
Options can be optionally passed in as the second parameter, as an object with a property ```goldwasher``` for goldwasher options, a property ```needle``` for needle options and a property ```retry``` for retry options. For instance:

```javascript
var options = {
    goldwasher: {
        selector: 'h1'
    },
    needle: {
        follow_max: 20
    },
    retry: {
        retries: 3
    }
}
```

Have a look at their respective doc pages for [goldwasher](https://www.npmjs.org/package/goldwasher), [needle](https://www.npmjs.org/package/needle) and [retry](https://www.npmjs.org/package/retry) for options available.

## Example
```javascript
var gn = require('goldwasher-needle');

gn('http://www.google.com', function(error, result) {
  console.log(result);
});
```

## Advanced example
```javascript
var gn = require('goldwasher-needle');

var url = 'http://www.google.com';
var options = {
  goldwasher: {
    selector: 'h1, h2, h3, h4, h5, h6, p'
  }
};

gn(url, options, function(error, result) {
  console.log(result);
});
```

CHANGELOG.md
# 2.0.0 (2015-10-09)

-- Change from plugin to module. require will now work differently. See example in README.

# 1.0.0 (2015-18-05)

## Features

-- First release.

# 0.0.1 (2015-18-05)

## Features

-- First commit.

test/goldwasher-needle.spec.js
// jscs:disable requireCamelCaseOrUpperCaseIdentifiers
'use strict';

var chai = require('chai');
chai.use(require('chai-things'));
var should = chai.should();
var http = require('http');
var R = require('ramda');
var gn = require('../lib/goldwasher-needle');
var defaultOptions = {
  retry: {
    minTimeout: 0,
    maxTimeout: 0
  }
};
var serverUrl = 'http://localhost:1337';
var url = serverUrl + '/200';
var server;

before(function(done) {
  var failures = 0;
  server = http.createServer(function(request, response) {
    var code = request.url.substring(1);

    if (request.url === '/200') {
      response.writeHead(code, {'Content-Type': 'text/plain'});
      response.end('<p>Hello World</p>');
    }

    if (request.url === '/301' || request.url === '/302') {
      response.writeHead(code, {'Content-Type': 'text/plain'});
      response.end('<p>Not found</p>');
    }

    if (request.url === '/404') {
      response.writeHead(code, {'Content-Type': 'text/plain'});
      response.end('<p>Not found</p>');
    }

    if (request.url === '/thirdtimeacharm') {
      if (failures < 2) {
        failures++;
        response.writeHead(500, {'Content-Type': 'text/plain'});
        response.end('<p>Server error!</p>');
      } else {
        response.writeHead(200, {'Content-Type': 'text/plain'});
        response.end('<p>Hello World</p>');
      }
    }
  }).listen(1337);

  server.on('listening', function() {
    done();
  });
});

after(function() {
  server.close();
});

describe('initialization', function() {

  it('loads without options', function(done) {
    gn(url, function(error, result, response) {
      done();
    });
  });

  it('loads with options', function(done) {
    var options = {
      goldwasher: {
        selector: 'p'
      }
    };

    gn(url, options, function(error, result, response) {
      done();
    });
  });

});

describe('running', function() {

  it('can request and goldwash a site', function(done) {
    gn(url, function(error, result, response) {
      should.not.exist(error);
      response.statusCode.should.equal(200);
      result.should.be.an('array');
      result.length.should.be.greaterThan(0);
      result[0].source.should.not.equal(null);
      done();
    });
  });

  it('can request and goldwash a site with options', function(done) {
    var options = {
      goldwasher: {
        selector: 'h1, h2, h3, h4, h5, h6, p'
      }
    };

    gn(url, options, function(error, result, response) {
      should.not.exist(error);
      response.statusCode.should.equal(200);
      result.should.be.an('array');
      result.length.should.be.greaterThan(0);
      result[0].source.should.not.equal(null);
      done();
    });
  });

  it('can request and return a raw site', function(done) {
    var options = {
      goldwasher: {
        output: 'raw'
      }
    };

    gn(url, options, function(error, result, response) {
      should.not.exist(error);
      response.statusCode.should.equal(200);
      result.should.be.a('string');
      done();
    });
  });

});

describe('failures', function() {

  it('handles errors from needle', function(done) {
    gn('foo', defaultOptions, function(error, result, response) {
      should.exist(error);
      done();
    });
  });

  it('throws on too many redirects', function(done) {
    var options = R.merge(defaultOptions, {
      needle: {
        follow_max: 0
      }
    });
    var url = serverUrl + '/301';

    gn(url, options, function(error, result, response) {
      should.exist(error);
      response.statusCode.should.be.within(301, 302);
      done();
    });
  });

  it('throws on 404', function(done) {
    var url = serverUrl + '/404';
    gn(url, defaultOptions, function(error, result, response) {
      should.exist(error);
      response.statusCode.should.equal(404);
      done();
    });
  });

  it('throws on disallowed header', function(done) {
    var options = R.merge(defaultOptions, {
      goldwasherNeedle: {
        disallowHeader: 'content-type'
      }
    });

    gn(url, options, function(error, result, response) {
      should.exist(error);
      done();
    });
  });
});

describe('retrying', function() {

  it('can retry on failure', function(done) {
    gn(
      serverUrl + '/thirdtimeacharm',
      defaultOptions,
      function(error, result) {
        should.not.exist(error);
        result.length.should.equal(1);
        result[0].text.should.equal('Hello World');
        done();
      });
  });

});

lib/goldwasher-needle.js
// jscs:disable requireCamelCaseOrUpperCaseIdentifiers
'use strict';

var R = require('ramda');
var goldwasher = require('goldwasher');
var needle = require('needle');
var retry = require('retry');

/**
 * Gets default options and merges with user options if any.
 * @param url
 * @param options
 * @returns {Object|*}
 */
var getDefaults = function(url, options) {
  var defaults = {
    goldwasher: {
      url: url
    },
    needle: {
      follow_max: 20
    },
    goldwasherNeedle: {
      disallowHeader: 'x-goldwasher-version'
    },
    retry: {
      retries: 5
    }
  };

  if (!options) {
    return defaults;
  }

  return {
    goldwasher: R.merge(defaults.goldwasher, options.goldwasher),
    needle: R.merge(defaults.needle, options.needle),
    retry: R.merge(defaults.retry, options.retry),
    goldwasherNeedle: R.merge(
      defaults.goldwasherNeedle,
      options.goldwasherNeedle
    )
  };
};

var run = function(url, options, callback) {
  needle.get(url, options.needle, function(error, response, body) {
    var result;

    if (error) {
      return callback(error);
    }

    if (response.statusCode === 301 || response.statusCode === 302) {
      return callback(
        new Error('Too many redirects. Increase in needle options.'),
        null,
        response
      );
    }

    if (response.statusCode !== 200) {
      return callback(
        new Error('Request failed. Status code: ' + response.statusCode),
        null,
        response
      );
    }

    if (response.headers[options.goldwasherNeedle.disallowHeader]) {
      return callback(
        new Error('Please do not call other goldwashers.'),
        null,
        response
      );
    }

    if (options.goldwasher.output === 'raw') {
      return callback(null, body, response);
    }

    result = goldwasher(body, options.goldwasher);

    return callback(null, result, response, body);
  });
};

/**
 * Use needle to get content of a site, pass to goldwasher and return result.
 * @param url
 * @param userOptions
 * @param callback
 */
module.exports = function(url, userOptions, callback) {
  var operation;
  var options;
  if (typeof (userOptions) === 'function') {
    callback = userOptions;
    options = getDefaults(url);
  } else {
    options = getDefaults(url, userOptions);
  }

  operation = retry.operation(options.retry);
  operation.attempt(function() {
    run(url, options, function(error, result, response, body) {
      if (operation.retry(error)) {
        return;
      }

      return callback(
        error ? operation.mainError() : null,
        result,
        response,
        body
      );
    });
  });
};